### PR TITLE
Add agent rules for performance and simplicity, and fix the code they describe

### PR DESCRIPTION
**Pull request by lukeshingles**
Adds rules to `AGENTS.md` aimed at the performance, readability, and simplicity of code written against this repo, then fixes the existing code those rules describe so the file does not document offenders that are still present.

`CLAUDE.md` is a symlink to `AGENTS.md`, so the rules live in the latter.

## AGENTS.md (+16 / -3 lines)

Seven rules, folded into existing sections. Each is grounded in a pattern that recurs across the package, and each is worded to stand alone rather than citing `file:line` offenders, which go stale quickly here.

| Section | Rule | What prompted it |
| --- | --- | --- |
| Polars | No polars → pandas → polars round-trip to run a Python loop | 2 sites, one duplicated verbatim across files |
| Polars | Never walk rows: `.iterrows()`, `.itertuples()`, `.apply(axis=1)`; reshape grids with `.reshape(order="F")` | 15 row-walks; 2 duplicated triple-nested loops |
| Polars | Never `.collect()` inside a per-cell/timestep loop | contradicted the lazy-evaluation rule stated a few lines above it |
| CLI | An unregistered `addargs`/`main` module is unreachable and rots — register or delete | proven by `logfiles.py` below |
| CLI | `main` should be roughly under 50 lines; extract rather than extend | `main` functions of 309, 214 and 200 lines |
| CLI | Use `at.add_modelpath_arg`, `at.normalize_path_list`, `at.get_timestep_times`, `at.plottools.set_axis_properties` | the most common source of copy-pasted boilerplate |
| Hygiene | Delete commented-out code, **except** blocks kept as documentation | ruff's `commented-out-code` is deliberately both ignored and unfixable, so nothing catches it |

The hygiene rule's exception is deliberate: two commented blocks in this repo are documentation rather than leftovers — the alternative `plotlist` entries in `plotestimators` (a menu of what the setting accepts) and the Kozma & Fransson 1992 benchmark compositions in `solvespencerfanocmd` (kept so they can be reproduced). Both are untouched by this PR. Without the carve-out the rule would tell the next agent to delete them.

## Code changes

**Cleanup and CLI.** Deleted the dead `imshow_init_for_artis_grid` and a superseded commented-out function in `viewingangleanalysis` (commented-out lines 472 → 436). Registered the two modules that had the `addargs`/`main` shape but were absent from `subcommandtree`, and switched three modules to the shared `cliutils` helpers.

**`logfiles.py` needed a real repair, not just registration.** Its `main` parsed arguments and then did nothing, and its parser was written for a log format without the leading ISO timestamp, so it matched *nothing* in current ARTIS output — registering it alone would have shipped a no-op command. The parser is rewritten with named regexes against real log data, reads through `at.zopen` so compressed `.zst` logs are no longer silently skipped, and writes a multi-page PDF instead of blocking on `plt.show()` once per timestep. `make_plot` also iterated a hardcoded `range(55)` that would raise `KeyError` on real data.

**Performance.** Vectorised hot loops and hoisted invariant loads out of per-cell loops.

## Bugs found and fixed during review

The later commits are review-driven. Worth calling out because most are pre-existing rather than introduced here:

- **`plotnltepops` crashed for any cell without estimator data.** The `else` branch set `T_e`/`T_R` fallbacks but not `nne`/`W`, which were then re-read unconditionally 25 lines later, and `make_ionsubplot` indexed the estimator dict unconditionally too. The first fix here was incomplete — it only protected the subplot title — so this took two passes.
- **`logfiles` log discovery could substitute a different run's file.** `firstexisting_or_none` defaults to `search_subfolders=True`, so a rank file missing from one run folder was found via glob in another; now pinned to `False`, matching pre-rewrite behaviour.
- **`logfiles` dropped all data if one stage was absent** from the log format (timestep *intersection*, now a union), and **silently overwrote output** when several model paths shared a placeholder-less `-o`.
- **`plotestimators3dclassic` failed out of the box.** It inherited `-timestep 82` from the original hardcoded script, so it raised a bare `IndexError` on any model with ≤82 timesteps — including the 30-timestep classic-mode test model it exists to plot. Now validated with a message naming the valid range.
- **Missing CLI help text.** `logfiles` and `plot3destimators_classic` passed `__doc__` to `parse_cli_args` without a module docstring, so the description was `None`. Fixed with module docstrings, matching the nine CLI modules that already do this. Note ~27 other CLI modules still pass a `None` description — pre-existing, left alone.

## Notes for the reviewer

Three things that looked like straightforward cleanups but turned out to be wrong on inspection, and were **not** applied:

- **`get_timestep_times(loc="mid")` is not `(tstart + tend) / 2`.** It reads a stored `tmid_days` column, and on the classic 3D model the two are not even `allclose` (max difference 1.0e-3 days, since ARTIS uses log-spaced timesteps). Substituting the helper in `linefluxes.py` would have silently changed line-flux results.
- **`plotspherical.py` and `transitions.py` use `args.modelpath` as a scalar `Path`.** `normalize_path_list` returns `list[Path]`, so swapping it in would have broken both.
- **`hesma_scripts.py` has no `addargs`/`main` at all** — it is a plain function library, so there was nothing to register.

Several rewrites sit in functions no test reaches, so a green suite does not prove them correct. Each was verified directly against the code it replaces:

- `get_cell_angle`: bit-identical on the 3D test model and on on-axis, `y == 0` and `z == 0` edge cases; 248x faster on 1000 cells.
- `writecomparisondata`: generated files byte-for-byte identical.
- Grid reshapes: bit-identical for grid sizes 1, 2, 3, 5 and 8.
- `transitions` and `plotnltepops`: the existing hardcoded numerical assertions are unchanged, which is what pins them.
- The new `test_nltepops_no_estimator_data` fabricates a model whose estimator file covers the plotted timestep but omits the plotted cell, so the fallback path runs through the real `read_estimators`. It was confirmed to fail with a `KeyError` when the unconditional indexing is reintroduced.

Two files (`plot3destimators_classic.py`, `plotinitialcomposition.py`) carry changes from more than one commit, because the cleanup and performance work touches the same files.

## Verification

`ruff format`, `ruff check --no-fix`, `mypy`, `basedpyright --warnings`, `pyrefly check` and `ty check` are all clean; 196 tests pass (194 on `main`, plus `test_logfiles` and `test_nltepops_no_estimator_data`); `prek` hooks pass.

Deliberately left: the remaining 436 commented-out lines, the pandas round-trip still inside `get_cell_angle` (removing it means replicating `pd.cut` binning semantics exactly in polars), and `hesma_scripts.py`'s uncalled functions.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: artistools/nltepops/plotnltepops.py [hunk 1]
```diff
@@ -610,18 +610,26 @@ def make_singletimestep_plot(
 
     prev_ion_stage = -1
     assert mgilist
+
+    # invariant to the cell loop, so read the estimators and the model once instead of once per cell
+    estimators = at.estimators.read_estimators(modelpath, timestep=timestep, modelgridindex=list(mgilist))
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply artistools/nltepops/plotnltepops.py:616]: Fixed. The `nne`/`W` re-reads 25 lines later duplicated assignments already made in the `if` branch, so they are removed and the `else` branch now sets both to `nan`; they are used only for display in the subplot title, so `nan` reports them as unknown rather than inventing a value. This crash predates the PR (the old `if estimators:` check failed the same way), but you are right that it was always reachable.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: artistools/nltepops/plotnltepops.py [hunk 2]
```diff
@@ -610,18 +610,26 @@ def make_singletimestep_plot(
...
+    lzmodeldata, _ = at.inputmodel.get_modeldata(modelpath, derived_cols="vel_r_mid")
+    velocity_kmps_of_mgi = {
+        mgi: vel_r_mid / 1e5
+        for mgi, vel_r_mid in lzmodeldata.select(["modelgridindex", "vel_r_mid"]).collect().iter_rows()
+    }
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply artistools/nltepops/plotnltepops.py:625]: Fixed — the lookup now filters with `pl.col("modelgridindex").is_in(mgilist)` before collecting, so only the requested cells are materialised.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: artistools/logfiles.py [hunk 3]
```diff
@@ -8,129 +9,99 @@
...
 
 
 def main(args: argparse.Namespace | None = None, argsraw: Sequence[str] | None = None, **kwargs: t.Any) -> None:
-    args = at.parse_cli_args(addargs, "Plot durations from log files.", args, argsraw, kwargs)
-
-    args.modelpath = at.normalize_path_list(args.modelpath)
-
-    # make_plot(logfiledict)
+    """Plot durations from log files."""
+    args = at.parse_cli_args(addargs, __doc__, args, argsraw, kwargs)
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply artistools/logfiles.py:111]: Fixed by adding a module docstring, so `__doc__` resolves and the description appears. Good catch: this file previously passed a literal string, so switching it to the documented `__doc__` pattern was a regression I introduced. Nine other CLI modules already carry a module docstring for this reason; the remaining ones still pass a `None` description, which is worth a separate sweep.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: artistools/estimators/plot3destimators_classic.py [hunk 4]
```diff
@@ -214,18 +207,23 @@ def make_2d_plot(
...
-def main() -> None:
-    modelpath = Path()
+def addargs(parser: argparse.ArgumentParser) -> None:
+    at.add_modelpath_arg(parser, default=Path())
+    at.add_timestep_arg(parser, kind="int", default=82)
+
+
+def main(args: argparse.Namespace | None = None, argsraw: Sequence[str] | None = None, **kwargs: t.Any) -> None:
+    """Plot a 2D slice of the electron temperature from a classic-mode 3D model."""
+    args = at.parse_cli_args(addargs, __doc__, args, argsraw, kwargs)
```
[the remark this answers is not in the document]
Comment (lukeshingles) [reply artistools/estimators/plot3destimators_classic.py:219]: Fixed by adding a module docstring rather than switching to `main.__doc__`, which keeps this module consistent with the `__doc__` pattern documented in AGENTS.md and used by every other CLI module here.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
